### PR TITLE
Adding the JART v1b memristor model for aihwkit

### PR DESCRIPTION
**Pull request by ZhenmingYu**
## Description

Support a new device type based on JART V1b memristor model and test scripts for that.
Details can be found at https://ieeexplore.ieee.org/document/10052010

## Details
The code was written for the old backend and some adaptations for the new version are needed.
Support for CUDA is strange, where sometimes the results are not transferred back to PyTorch.
Cycle-to-cycle variability is currently not supported for CUDA, as we need fields to save more device-specific parameters.

### CODE REVIEW

**Comment by maljoras** [comment]
@ZhenmingYu, actually I cannot edit your branch anyway, so I will first review it, so that you can make changes

**Comment by ZhenmingYu** [comment]
@maljoras Thanks a lot for the kind words! I couldn't have done this without your support last year.
So I tried to invite you as a collaborator under my fork, this should enable you to edit the branch.
My plan is to keep the JART_v1b_device branch as is and improve and merge the master branch.

I'm also wondering whether there can be more than 6 device-specific parameters in version 0.7 now with the new backend, which could be helpful with c2c noises.

Thanks a lot for the help!
Zhenming

**Review by maljoras — CHANGES_REQUESTED** [review]

**Comment by maljoras** [comment]
We could add some more device-to-device parameters into the kernels. like another par_4 (to have 4 values per synapse). Would this be enough? So you need to modify  them or are they constant?

**Comment by ZhenmingYu** [comment]
> 
We have implemented a random walk based c2c noise that acts on 4 parameters, starting from their d2d values.
The parameters themselves are already included in the 6 device-to-device parameters, and they each need 2 boundaries to prevent random walks from going too far.

So there are 8 constants in total, that are equal to Par_d2d*(1+-10%), calculated after d2d noises being applied.
Adding 2 par4s would then be enough.

File: src/rpucuda/cuda/rpucuda_JART_v1b_device.cu [hunk 1]
```diff
@@ -0,0 +1,576 @@
+/**
+ * (C) Copyright 2022 Forschungszentrum Jülich GmbH, Zhenming Yu. All Rights reserved.
```
[the remark this answers is not in the document]
Comment (ZhenmingYu) [reply src/rpucuda/cuda/rpucuda_JART_v1b_device.cu:2]: Sure, I asked the Forschungszentrum Jülich legal department before about this, and apparently, they have no rules on this end.
This was a suggested line that they mentioned, but I think we can have flexibility on this.

File: src/rpucuda/cuda/rpucuda_JART_v1b_device.cu [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+          // NOTE: dWamin = dWa_f = dWa*(sqrt(1.0-pow(gamma,2.0))-(gamma*M_PI)/2+gamma*asin(gamma)) = dWa_mean - dWa_difference
+          // NOTE: dWamax = dWa_r = dWa*(sqrt(1.0-pow(gamma,2.0))+(gamma*M_PI)/2+gamma*asin(gamma)) = dWa_mean + dWa_difference
+          T dWa_mean = dWa*(sqrt(1.0-pow(gamma,2.0))+gamma*asin(gamma));
+          T dWa_difference = dWa*((gamma*M_PI)/2.0);
+
+          T denominator = PHYSICAL_PARAMETER_kb_over_e*Treal;
+
+          T c_v0 = (Nplug+Ndisc_double)/2.0;
+          T F_limit = 1.0-pow((device_specific_Ndiscmin_cuda/Ndisc_double),10.0);
+          T dNdt = -(c_v0*a_ny0*F_limit*(exp(-(dWa_mean - dWa_difference)/denominator)-exp(-(dWa_mean + dWa_difference)/denominator)))/device_specific_ldisc_cuda;
```
[the remark this answers is not in the document]
Comment (ZhenmingYu) [reply src/rpucuda/cuda/rpucuda_JART_v1b_device.cu:254]: So dt here is only calculated within a pulse.
If an update is not needed, no voltage is applied, and dNdt during that physical time will be zero, thus ignored in the code.

File: src/rpucuda/cuda/rpucuda_JART_v1b_device.cu [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+//   map_weight_to_Ndisc<T>(
+//       this->context_, weights, Ndisc, this->size_,
+      // par.current_min, par.w_min, par.weight_to_current_ratio,
+      // par.g_read, par.h_read, par.j_0, par.k0, par.Ndiscmin);
+
+// }
+
+template <typename T> void JARTv1bRPUDeviceCuda<T>::clipWeights(T *weights, T clip) {
+
+  RPU::math::elemsat<T>(this->context_, weights, this->size_, this->dev_4params_->getData());
```
[the remark this answers is not in the document]
Comment (ZhenmingYu) [reply src/rpucuda/cuda/rpucuda_JART_v1b_device.cu:453]: Yes, they are, I re-used the getMaxBound() function to get the Max Bound for Oxygen vacancy concentration, Ndisc, which is stored in the hidden persistent weight.

File: JART_v1b_tests/00_basic_test_v1b.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        wandb.init(project=project_name, group="Linear Regression", job_type=job_type)
+        wandb.config.update(config_dictionary)
+
+    # Prepare the datasets (input and expected output).
+    slope = -0.5
+    x = Tensor([[0.0], [1.0], [2.0], [3.0], [4.0]])
+    y = Tensor([[0.0], [slope], [2*slope], [3*slope], [4*slope]])
+
+    # Define a single-layer network.
+    rpu_config = SingleRPUConfig(device=JARTv1bDevice(w_max=config_dictionary["w_max"],
```
[the remark this answers is not in the document]
Comment (ZhenmingYu) [reply JART_v1b_tests/00_basic_test_v1b.py:80]: Did something similar,a yaml_loader.py file that holds the configuration stufff.
I also put stuff like USE_bias in there, so putting it in JARTv1bDevice might not be ideal.

File: src/rpucuda/cuda/rpucuda_JART_v1b_device.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (maljoras) [reply src/rpucuda/cuda/rpucuda_JART_v1b_device.cu:254]: I see. That seems fine then

**Comment by maljoras** [comment]
OK, I could add 2 par_4 that are both writeable. the current par_2 and par_4 are considered constant. Then you have 8 more writable parameters per synapse, in addition to the 6 constant (par_2, par_4) and the weight and "hidden" weight.

File: src/rpucuda/cuda/rpucuda_JART_v1b_device.cu [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+        //                            ldisc_ctoc_upper_bound, ldisc_ctoc_lower_bound, rdisc_ctoc_upper_bound, rdisc_ctoc_lower_bound);
+        Ndisc_double = MAX(Ndisc_double, min_bound);
+        w = map_Ndisc_to_weight(Ndisc_double, current_min, w_min, current_to_weight_ratio, g_read, h_read, j_0, k0, Original_Ndiscmin);
+        Ndisc = Ndisc_double;
+      }
+    }
+    // TODO: BUG: Removing this delay or the print line will result in PyTorch not receving the updated weights.
+    // printf("w after update %.20f\n", apparent_weight);
+    uint32_t ns = 1;
+    __nanosleep(ns);
```
[the remark this answers is not in the document]
Comment (ZhenmingYu) [reply src/rpucuda/cuda/rpucuda_JART_v1b_device.cu]: Tested, __syncthreads() did work.
However, the performance of the CUDA version seems to be quite slow compared to the GPU version.

File: src/rpucuda/cuda/rpucuda_JART_v1b_device.cu [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+    T device_specific_Ndisc_max_bound_cuda = par_4.z;                          // [2]
+
+    /* NOTE: These values do random walks,
+             use refference to change the recorded values.
+    */ 
+    T &device_specific_Ndiscmin_cuda = par_4.y; // [1]
+    T &device_specific_Ndiscmax_cuda = par_4.w; // [3]
+    T &device_specific_ldisc_cuda = par_2.x; // [0]
+    T &device_specific_A_cuda = par_2.y; // [1]
+
```
[the remark this answers is not in the document]
Comment (ZhenmingYu) [reply src/rpucuda/cuda/rpucuda_JART_v1b_device.cu:180]: True. If I comment out the c2c variations, the torch weight is updated.
I changed these back to constant values and will wait for the 2 extra writable par4s.

**Comment by ZhenmingYu** [comment]
> 

That would be great. Thanks a lot!
Although I've discovered that the CUDA version runs pretty slow compared to the CPU version some how.

**Comment by maljoras** [comment]
The GPU kernel currently seems quite complex one could potentially improve the performance. However, for small batch sizes and small weights the CPU version always will be faster, since there is quite an overhead of memory transfer etc in the CUDA kernel. If the compute amount is too small, that part will dominate. If you increase the batch size and weight size GPU should become much faster than CPU.

File: src/rpucuda/cuda/rpucuda_JART_v1b_device.cu [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+          T dWa_mean = dWa*(sqrt(1.0-pow(gamma,2.0))+gamma*asin(gamma));
+          T dWa_difference = dWa*((gamma*M_PI)/2.0);
+
+          T denominator = PHYSICAL_PARAMETER_kb_over_e*Treal;
+
+          T c_v0 = (Nplug+Ndisc_double)/2.0;
+          T F_limit = 1.0-pow((Ndisc_double/device_specific_Ndiscmax_cuda),10.0);
+          T dNdt = -(c_v0*a_ny0*F_limit*(exp(-(dWa_mean - dWa_difference)/denominator)-exp(-(dWa_mean + dWa_difference)/denominator)))/device_specific_ldisc_cuda;
+
+          Ndisc_double = Ndisc_double + dNdt*base_time_step;
```
Comment (maljoras) [inline src/rpucuda/cuda/rpucuda_JART_v1b_device.cu:223]: To improve performance, you should put all calculations that are repeated in the loop out side of the loop. In particular if the involve only constant numbers. There are all not necessary and could even be done once on CPU.  

Also note that "pow" and "asin" and "exp" are quite expensive. They could be replaced with the fast __expf etc. or x*x if pow(x, 2).

**Comment by ZhenmingYu** [comment]
Thanks for the suggestions, it is much faster now.

**Comment by maljoras** [comment]
moved to #583  since I cannot edit this PR

**Pull request closed without merging**